### ISSUE TITLE
Understanding the impact of GetTensors and SyncTensors on memory usage

### ISSUE THREAD

**Issue by ymwangg** [pytorch/xla#3102]
We recently noticed an interesting behavior about the impact of `GetTensors` and `SyncTensors` on memory usage. We found inserting a `print(loss)` at the end of each training loop or inserting an extra `mark_step` just before `optimizer.step` can dramatically improve the maximum supported batch sizes. For example, on Tesla V100, the maximum batch size the following script can run is 17 with `--print_loss` or `--extra_mark_step`. However, the maximum supported batch size reduced to 9 without these two options.
```python
from transformers import BertForMaskedLM
from torch.optim import AdamW
import torch
import torch_xla.core.xla_model as xm
import argparse
import time

def train(model, optimizer, batch_size, num_steps, print_loss=False, extra_mark_step=False):
    device = xm.xla_device()
    input_ids = torch.ones((batch_size, 512)).to(torch.int64).to(device)
    attention_mask = torch.ones((batch_size, 512)).to(torch.int64).to(device)
    labels = torch.ones((batch_size, 512)).to(torch.int64).to(device)
    print("batch_size={}".format(batch_size))
    t0 = time.time()
    for step in range(num_steps):
        optimizer.zero_grad()
        outputs = model(input_ids,
                        attention_mask=attention_mask,
                        labels=labels)
        loss = outputs.loss
        loss.backward()
        if extra_mark_step:
            xm.mark_step()
        xm.optimizer_step(optimizer)
        xm.mark_step()
        print("step={}".format(step))
        if print_loss:
            print("loss={}".format(loss))
    t1 = time.time()
    print("dt={}".format(t1-t0))

if __name__ == "__main__":
    arg_parser = argparse.ArgumentParser()
    arg_parser.add_argument("--batch_size", type=int, default=12)
    arg_parser.add_argument("--print_loss", action="store_true", default=False)
    arg_parser.add_argument("--extra_mark_step", action="store_true", default=False)
    args = arg_parser.parse_args()
    batch_size = args.batch_size
    print_loss = args.print_loss
    extra_mark_step = args.extra_mark_step
    device = xm.xla_device()
    model = BertForMaskedLM.from_pretrained("bert-base-uncased")
    model.to(device)
    model.train()
    optimizer = AdamW(model.parameters(), lr=5e-5)
    train(model, optimizer, batch_size, 100, print_loss, extra_mark_step)
```
Why an extra `--print_loss` or `--extra_mark_step` has so much memory impact? Is it possible to support the same maximum batch size without adding these two options?
We noticed XLA aliasing for in-place update of gradients may have such large memory impact but not sure if this is relevant to this problem.

**Comment by ymwangg** [comment]
After some digging, I suspect this may be due to asynchronous execution in which `step i` and `step 1+1` consume GPU memory at the same time. I tested the effect of `time.sleep` and a manually-added `xm.barrier` function that performs `locker->Unlock(); locker->Unlock();` of `torch_xla::DeviceLocker`.
I tested different execution orders inside the training loop, and the following sequences can run with max batch size = 17:
- `forward-backward -> optimizer.step-> xm.mark_step` with `XLA_SYNC_WAIT=1`
- `forward-backward -> optimizer.step -> xm.mark_step -> xm.barrier`
- `forward-backward -> xm.barrier() -> optimizer.step -> xm.mark_step`
- `forward-backward -> optimizer.step -> xm.mark_step -> time.sleep(0.01)`

However the following cases can only run with max batch size = 9:
- `forward-backward -> optimizer.step -> xm.mark_step`
- `forward-backward -> optimizer.step -> xm.barrier() -> xm.mark_step`
- `forward-backward -> optimizer.step -> xm.mark_step -> time.sleep(0.001)`

Given the maximum batch size is almost 2x when `time.sleep(0.01)` is added, the extra memory may be due to the overlapped memory usage between `step i` and `step i+1`. It looks like the device lock is set [here](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L1142) and is released after this [function](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L1311) is done. @JackCaoG  Is it possible that other places not guarded by the device lock also consumes the GPU memory? Or is this an intended behavior?

**Comment by JackCaoG** [comment]
Can you check the `transferToServer` counter(it should also have sizes) before and after the `F.adamw`? There can be at most 1 execution happens in the same time(due to the device lock) but it is possible that `F.adamw` created a lot host->device upload which also consumes device memory. The transferToServer is in [here](https://github.com/pytorch/xla/blob/master/third_party/xla_client/xrt_computation_client.cc#L352).

**Comment by ymwangg** [comment]
@JackCaoG Thanks a lot, this is very helpful. After adding a counter in [TensorToXlaData](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor_util.h#L32), I found `F.adamw` triggered `TensorToXlaData` 2 times while SGD does not trigger it at all after the first iteration. Furthermore, these two `TensorToXlaData` func calls are due to these two [variables](https://github.com/pytorch/pytorch/blob/master/torch/optim/_functional.py#L127-L128) that change every iteration. If I make these two variables constant like `bias_correction1 = 1; bias_correction2 = 1;`, this memory overhead is gone. I compared the IR dump before and after making these two variables constant, however there are no differences. Is it possible that `a.mul_(at::Scalar b)` and `a.div_(at::Scalar b)` functions are broadcasting `b` to the shape of `a` before uploading to devices? It feels like this memory overhead should be avoided.

**Comment by JackCaoG** [comment]
```
I compared the IR dump before and after making these two variables constant, however there are no differences.
```

I think this part is expected. The IR value for two variable would just be `xla::device_data()`. The difference is that when the value of this device data changed, we need to reupload them. I think we have a cache for `device_data`, when you set those values to constant, it is likely you hit a `device_data` cache so no need to reupload.

For you second question, we actually just upload a scalar and then do a expand(broadcast) to it. You can check the code in [here](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor_methods.cpp#L1799) and [here](https://github.com/pytorch/xla/blob/132255a343bca2070d9fe27ff258092c6508636d/torch_xla/csrc/tensor.cpp#L742)
